app/controllers/projects_controller.rb

```diff
@@ -333,6 +333,46 @@ def create_reference
     end
   end
 
+  #Fang retuen the data for the progress of a project.
+  def progress
+    id = params[:id]
+    project = Project.find(id)
+    status = project.miga.datasets.map(&:status)
+    incomplete = status.count(:incomplete)
+    total = status.size
+    inactive = status.count(:inactive)
+    complete = (total - incomplete) - inactive
+    p_incomplete = (incomplete.to_f * 100.0)/total
+    p_complete = 100.0 - p_incomplete
+
+    p_complete = p_complete.round(2)
+   
+    #reading outputfile
+    f = File.expand_path("daemon/MiGA\:#{project.miga.name}.output", project.miga.path)
+    last_line = `tail -n 1 #{f}`
+    logger.info("last 1 : " + last_line)
+    last_five_lines = `tail -n 5 #{f}`
+    logger.info ("last 5 : " + last_five_lines)
```

Comment:
That shell access wasn't safe, and would easily break (e.g., with a project path including spaces)
```suggestion
    last_five_lines = File.readlines(f)[-5, 5]
    last_line = last_five_lines.last
    logger.info("last 1 : " + last_line)
    logger.info ("last 5 : " + last_five_lines)
```